File: web_api/routes/knowledge_center.py
from typing import Any

from fastapi import APIRouter, Depends, HTTPException

from web_api.errors import ApiError, api_error_response
from web_api.deps import get_knowledge_center_service
from web_api.schemas.knowledge_center import (
    IndexRunRequest,
    IndexRunResponse,
    KnowledgeListResponse,
    ProductOverrideUpdate,
    ProductPublishResponse,
    SopCreate,
    SopPublishResponse,
    SopUpdate,
    VersionChunksResponse,
)
from web_api.services.knowledge_center_service import KnowledgeCenterService, empty_product_override
from web_api.services.shop_identity import assert_real_shop_id_bound


router = APIRouter(prefix="/knowledge", tags=["knowledge"])


@router.get("/sop", response_model=KnowledgeListResponse)
def list_sop(
    shop_id: str | None = None,
    domain: str | None = None,
    status: str | None = None,
    service: KnowledgeCenterService = Depends(get_knowledge_center_service),
) -> dict[str, Any]:
    items = service.list_sop(shop_id=shop_id, domain=domain, status=status)
    return {"items": items, "total": len(items)}


@router.post("/sop")
def create_sop(payload: SopCreate, service: KnowledgeCenterService = Depends(get_knowledge_center_service)) -> dict[str, Any]:
    return service.create_sop(payload.shop_id, payload.domain, payload.title, payload.content)


@router.get("/sop/{sop_id}")
def get_sop(sop_id: int, service: KnowledgeCenterService = Depends(get_knowledge_center_service)) -> dict[str, Any]:
    try:
        return service.get_sop(sop_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.put("/sop/{sop_id}")
def update_sop(sop_id: int, payload: SopUpdate, service: KnowledgeCenterService = Depends(get_knowledge_center_service)) -> dict[str, Any]:
    try:
        current = service.get_sop(sop_id)
        if payload.expected_content_hash and payload.expected_content_hash != current["content_hash"]:
            raise HTTPException(
                status_code=409,
                detail={"status": "conflict", "current_content_hash": current["content_hash"]},
            )
        return service.update_sop(sop_id, title=payload.title, content=payload.content)
    except HTTPException:
        raise
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.delete("/sop/{sop_id}")
def archive_sop(sop_id: int, service: KnowledgeCenterService = Depends(get_knowledge_center_service)) -> dict[str, Any]:
    try:
        return service.archive_sop(sop_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.post("/sop/{sop_id}/publish", response_model=SopPublishResponse)
def publish_sop(sop_id: int, service: KnowledgeCenterService = Depends(get_knowledge_center_service)) -> dict[str, Any]:
    try:
        sop = service.get_sop(sop_id)
        assert_real_shop_id_bound(sop["shop_id"])
        result = service.publish_sop(sop["shop_id"], sop_id)
        return {"sop": sop, "version": result["version"], "index_job": result["job"]}
    except ApiError as exc:
        return api_error_response(exc)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.get("/products/{goods_id}/overrides")
def get_product_override(
    goods_id: str,
    shop_id: str,
    service: KnowledgeCenterService = Depends(get_knowledge_center_service),
) -> dict[str, Any]:
    override = service.get_override(shop_id, goods_id)
    return override or empty_product_override(shop_id, goods_id)


@router.put("/products/{goods_id}/overrides")
def upsert_product_override(
    goods_id: str,
    shop_id: str,
    payload: ProductOverrideUpdate,
    service: KnowledgeCenterService = Depends(get_knowledge_center_service),
) -> dict[str, Any]:
    current = service.get_override(shop_id, goods_id)
    if payload.expected_content_hash and payload.expected_content_hash != (current or {}).get("content_hash", ""):
        raise HTTPException(
            status_code=409,
            detail={"status": "conflict", "current_content_hash": (current or {}).get("content_hash", "")},
        )
    data = payload.model_dump(exclude={"expected_content_hash"})
    return service.upsert_override(shop_id, goods_id, **data)


@router.get("/products/{goods_id}/effective")
def get_effective_product(
    goods_id: str,
    shop_id: str,
    service: KnowledgeCenterService = Depends(get_knowledge_center_service),
) -> dict[str, Any]:
    try:
        return service.build_effective_product_knowledge(shop_id, goods_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.post("/products/{goods_id}/publish", response_model=ProductPublishResponse)
def publish_product(
    goods_id: str,
    shop_id: str,
    service: KnowledgeCenterService = Depends(get_knowledge_center_service),
) -> dict[str, Any]:
    try:
        assert_real_shop_id_bound(shop_id)
        result = service.publish_product(shop_id, goods_id)
        return {"effective": result["effective"], "version": result["version"], "index_job": result["job"]}
    except ApiError as exc:
        return api_error_response(exc)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.get("/versions", response_model=KnowledgeListResponse)
def list_versions(
    shop_id: str | None = None,
    source_type: str | None = None,
    source_id: str | None = None,
    domain: str | None = None,
    status: str | None = None,
    is_active: bool | None = None,
    service: KnowledgeCenterService = Depends(get_knowledge_center_service),
) -> dict[str, Any]:
    items = service.list_versions(
        shop_id=shop_id,
        source_type=source_type,
        source_id=source_id,
        domain=domain,
        status=status,
        is_active=is_active,
    )
    return {"items": items, "total": len(items)}


@router.get("/versions/{version_id}")
def get_version(version_id: int, service: KnowledgeCenterService = Depends(get_knowledge_center_service)) -> dict[str, Any]:
    try:
        return service.get_version(version_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.get("/versions/{version_id}/chunks", response_model=VersionChunksResponse)
def get_version_chunks(
    version_id: int,
    limit: int = 20,
    service: KnowledgeCenterService = Depends(get_knowledge_center_service),
) -> dict[str, Any]:
    try:
        return service.list_version_chunks(version_id, limit=limit)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.get("/index-jobs", response_model=KnowledgeListResponse)
def list_index_jobs(
    shop_id: str | None = None,
    status: str | None = None,
    source_type: str | None = None,
    source_id: str | None = None,
    version_id: int | None = None,
    service: KnowledgeCenterService = Depends(get_knowledge_center_service),
) -> dict[str, Any]:
    items = service.list_index_jobs(
        shop_id=shop_id,
        status=status,
        source_type=source_type,
        source_id=source_id,
        version_id=version_id,
    )
    return {"items": items, "total": len(items)}


@router.get("/index-jobs/{job_id}")
def get_index_job(job_id: int, service: KnowledgeCenterService = Depends(get_knowledge_center_service)) -> dict[str, Any]:
    try:
        return service.get_index_job(job_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.post("/index-jobs/{job_id}/run", response_model=IndexRunResponse)
def run_index_job(
    job_id: int,
    payload: IndexRunRequest | None = None,
    mode: str | None = None,
    embedding_provider: str | None = None,
    vector_store: str | None = None,
    service: KnowledgeCenterService = Depends(get_knowledge_center_service),
) -> dict[str, Any]:
    try:
        options = payload or IndexRunRequest()
        return service.run_index_job(
            job_id,
            mode=mode or options.mode,
            embedding_provider=embedding_provider or options.embedding_provider,
            vector_store_provider=vector_store or options.vector_store,
        )
    except ValueError as exc:
        raise HTTPException(status_code=409, detail={"status": "not_runnable", "error_type": str(exc).split(":")[0]}) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=500, detail={"status": "failed", "error_type": str(exc).split(":")[0]}) from exc
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc


@router.post("/index-jobs/{job_id}/retry")
def retry_index_job(job_id: int, service: KnowledgeCenterService = Depends(get_knowledge_center_service)) -> dict[str, Any]:
    try:
        return service.retry_job(job_id)
    except ValueError as exc:
        raise HTTPException(status_code=409, detail={"status": "not_retryable", "error_type": str(exc).split(":")[0]}) from exc
    except KeyError as exc:
        raise HTTPException(status_code=404, detail={"status": "not_found"}) from exc

File: web_frontend/src/api/knowledgeCenter.ts
import { apiDelete, apiGet, apiPost, apiPut } from "./client";

export type KnowledgeList<T> = {
  items: T[];
  total: number;
};

export type SopRecord = {
  id: number;
  shop_id: string;
  domain: string;
  title: string;
  content: string;
  status: string;
  content_hash: string;
  created_by: string;
  updated_by: string;
  created_at: string;
  updated_at: string;
};

export type KnowledgeVersion = {
  id: number;
  shop_id: string;
  source_type: string;
  source_id: string;
  domain: string;
  version: string;
  content_hash: string;
  snapshot_json: string;
  status: string;
  is_active: number | boolean;
  index_job_id: number | null;
  created_by: string;
  created_at: string;
  indexed_at: string | null;
  activated_at: string | null;
  retired_at: string | null;
  error_summary: string | null;
};

export type KnowledgeIndexJob = {
  id: number;
  shop_id: string;
  version_id: number;
  source_type: string;
  source_id: string;
  domain: string;
  version: string;
  status: string;
  chunk_count: number;
  embedded_count: number;
  indexed_count: number;
  error_summary: string | null;
  retry_count: number;
  index_run_id: string;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
};

export type SopCreatePayload = {
  shop_id: string;
  domain: string;
  title: string;
  content: string;
};

export type SopUpdatePayload = {
  title?: string;
  content?: string;
  expected_content_hash?: string;
};

export type SopPublishResponse = {
  sop: SopRecord;
  version: KnowledgeVersion;
  index_job: KnowledgeIndexJob;
};

export type ProductOverride = {
  id?: number | null;
  shop_id: string;
  goods_id: string;
  goods_name: string | null;
  usage_override: string | null;
  ingredients_override: string | null;
  warnings_override: string | null;
  shelf_life_override: string | null;
  manual_notes: string | null;
  specs_override: string | null;
  price_note_override: string | null;
  status: string;
  content_hash: string;
  created_by?: string | null;
  updated_by?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
};

export type ProductOverridePayload = Partial<
  Pick<
    ProductOverride,
    | "goods_name"
    | "usage_override"
    | "ingredients_override"
    | "warnings_override"
    | "shelf_life_override"
    | "manual_notes"
    | "specs_override"
    | "price_note_override"
  >
> & {
  expected_content_hash?: string;
};

export type EffectiveField = {
  value: string | string[] | number | null;
  source: "raw" | "manual_override" | "missing" | string;
};

export type EffectiveProduct = {
  shop_id: string;
  goods_id: string;
  goods_name: string;
  fields: Record<string, EffectiveField>;
  raw: Record<string, unknown>;
  override: ProductOverride | Record<string, unknown>;
};

export type ProductPublishResponse = {
  effective: EffectiveProduct;
  version: KnowledgeVersion;
  index_job: KnowledgeIndexJob;
};

export type IndexRunResponse = {
  version: KnowledgeVersion;
  index_job: KnowledgeIndexJob;
  index_mode?: string;
  error_type?: string;
};

export type IndexRunPayload = {
  mode?: "real";
  embedding_provider?: "doubao" | "ark" | "ollama";
  vector_store?: "pgvector";
};

export type VersionChunk = {
  chunk_id: string;
  version_id: number | null;
  shop_id: string;
  domain: string;
  source_type: string;
  source_id: string;
  version: string;
  content: string;
  content_hash: string;
  chunk_hash: string;
  index_run_id: string;
  metadata: Record<string, unknown>;
  created_at: string;
};

export type VersionChunksResponse = {
  version_id: number;
  source_type: string;
  source_id: string;
  domain: string;
  version: string;
  chunks: VersionChunk[];
  warning: string | null;
};

export type SopQuery = {
  shop_id?: string;
  domain?: string;
  status?: string;
};

export type VersionQuery = {
  shop_id?: string;
  source_type?: string;
  source_id?: string;
  domain?: string;
  status?: string;
  is_active?: boolean | string;
};

export type IndexJobQuery = {
  shop_id?: string;
  status?: string;
  source_type?: string;
  source_id?: string;
  version_id?: number | string;
};

function buildQuery(params: Record<string, string | number | boolean | undefined | null>): string {
  const searchParams = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== "") {
      searchParams.set(key, String(value));
    }
  });
  const query = searchParams.toString();
  return query ? `?${query}` : "";
}

export function listSop(params: SopQuery = {}) {
  return apiGet<KnowledgeList<SopRecord>>(`/api/knowledge/sop${buildQuery(params)}`);
}

export function createSop(payload: SopCreatePayload) {
  return apiPost<SopRecord>("/api/knowledge/sop", payload);
}

export function getSop(id: number) {
  return apiGet<SopRecord>(`/api/knowledge/sop/${id}`);
}

export function updateSop(id: number, payload: SopUpdatePayload) {
  return apiPut<SopRecord>(`/api/knowledge/sop/${id}`, payload);
}

export function archiveSop(id: number) {
  return apiDelete<SopRecord>(`/api/knowledge/sop/${id}`);
}

export function publishSop(id: number) {
  return apiPost<SopPublishResponse>(`/api/knowledge/sop/${id}/publish`, {});
}

export function getProductOverride(goodsId: string, shopId: string) {
  return apiGet<ProductOverride>(
    `/api/knowledge/products/${encodeURIComponent(goodsId)}/overrides${buildQuery({ shop_id: shopId })}`
  );
}

export function saveProductOverride(goodsId: string, shopId: string, payload: ProductOverridePayload) {
  return apiPut<ProductOverride>(
    `/api/knowledge/products/${encodeURIComponent(goodsId)}/overrides${buildQuery({ shop_id: shopId })}`,
    payload
  );
}

export function getEffectiveProduct(goodsId: string, shopId: string) {
  return apiGet<EffectiveProduct>(
    `/api/knowledge/products/${encodeURIComponent(goodsId)}/effective${buildQuery({ shop_id: shopId })}`
  );
}

export function publishProduct(goodsId: string, shopId: string) {
  return apiPost<ProductPublishResponse>(
    `/api/knowledge/products/${encodeURIComponent(goodsId)}/publish${buildQuery({ shop_id: shopId })}`,
    {}
  );
}

export function listVersions(params: VersionQuery = {}) {
  return apiGet<KnowledgeList<KnowledgeVersion>>(`/api/knowledge/versions${buildQuery(params)}`);
}

export function getVersion(id: number) {
  return apiGet<KnowledgeVersion>(`/api/knowledge/versions/${id}`);
}

export function getVersionChunks(id: number, limit = 20) {
  return apiGet<VersionChunksResponse>(`/api/knowledge/versions/${id}/chunks${buildQuery({ limit })}`);
}

export function listIndexJobs(params: IndexJobQuery = {}) {
  return apiGet<KnowledgeList<KnowledgeIndexJob>>(`/api/knowledge/index-jobs${buildQuery(params)}`);
}

export function getIndexJob(id: number) {
  return apiGet<KnowledgeIndexJob>(`/api/knowledge/index-jobs/${id}`);
}

export function runIndexJob(
  id: number,
  payload: IndexRunPayload = { mode: "real", embedding_provider: "doubao", vector_store: "pgvector" }
) {
  return apiPost<IndexRunResponse>(`/api/knowledge/index-jobs/${id}/run`, {
    mode: "real",
    embedding_provider: "doubao",
    vector_store: "pgvector",
    ...payload,
  });
}

export function retryIndexJob(id: number) {
  return apiPost<KnowledgeIndexJob>(`/api/knowledge/index-jobs/${id}/retry`, {});
}
